### ISSUE TITLE
Error in numerical integration code since a657fa0

### ISSUE THREAD

**Issue by ladc** [LuaJIT/LuaJIT#193]
Since commit a657fa0, one of my numerical integration routines produces NaN values. To reproduce:

`wget https://github.com/Climdyn/MAOOAM/archive/master.zip && unzip master.zip && cd MAOOAM-master/lua && luajit maooam.lua`

The assertion that none of the variables should be NaN often fails for me (`NaN encountered; the system is unstable. Please modify your initial conditions, forcing or coupling parameter.`). I also found it to fail more reliably when the binary has a longer filename (`luajit-2.1.0-beta2` is more likely to fail than just `luajit`), which indicates a memory corruption issue.

Using x86_64 (Ubuntu 16.04), GCC version 5.3.1.

I still need to reduce this to a minimal test case, sorry.

**Comment by ladc** [comment]
When testing this on 32-bit x86, I get an error in a different location:
`luajit-2.1.0-beta2: ./inprod_analytic.lua:96: ./inprod_analytic.lua:77: attempt to perform arithmetic on local 'sign' (a function value)`

The error does not occur with `-joff`.

**Comment by starius** [comment]
I have random failures in [luacheck](https://github.com/mpeterv/luacheck) and my own code with https://github.com/LuaJIT/LuaJIT/commit/287a5347cfe452d44748327fb7c27f6ce57f5dc2

How to reproduce the error in luacheck (sorry for not minimizing it):

```
mkdir here
wget https://raw.githubusercontent.com/mpeterv/hererocks/13e45c9db0293ac59ffb3756c0a883c7221e9295/hererocks.py
python hererocks.py here --luajit 2.1 --luarocks 2.2.2
./here/bin/luarocks install luacheck
git clone https://github.com/npge/lua-npge
./here/bin/luacheck `find lua-npge/src -name '*.lua'`
```

Possible results: normal operation, segfault or the following error message:

```
Critical error: Luacheck 0.15.1 bug (please report at github.com/mpeterv/luacheck/issues):
.../here/share/lua/5.1/luacheck/linearize.lua:160: attempt to index field 'location' (a nil value)
stack traceback:
        .../here/share/lua/5.1/luacheck/utils.lua:162: in function '__index'
        .../here/share/lua/5.1/luacheck/linearize.lua:160: in function 'leave_scope'
        .../here/share/lua/5.1/luacheck/linearize.lua:661: in function 'build_line'
        .../here/share/lua/5.1/luacheck/linearize.lua:673: in function 'scanner'
        .../here/share/lua/5.1/luacheck/linearize.lua:479: in function 'scan_expr'
        .../here/share/lua/5.1/luacheck/linearize.lua:442: in function <.../here/share/lua/5.1/luacheck/linearize.lua:438>
        .../here/share/lua/5.1/luacheck/linearize.lua:281: in function 'emit_stmt'
        .../here/share/lua/5.1/luacheck/linearize.lua:286: in function 'emit_stmts'
        .../here/share/lua/5.1/luacheck/linearize.lua:292: in function 'emit_block'
        .../here/share/lua/5.1/luacheck/linearize.lua:377: in function <.../here/share/lua/5.1/luacheck/linearize.lua:368>
        .../here/share/lua/5.1/luacheck/linearize.lua:281: in function 'emit_stmt'
        ...
        .../here/share/lua/5.1/luacheck/check.lua:209: in function 'func'
        .../here/share/lua/5.1/luacheck/utils.lua:261: in function 'map'
        .../here/share/lua/5.1/luacheck/main.lua:375: in function 'get_new_reports'
        .../here/share/lua/5.1/luacheck/main.lua:420: in function 'get_reports'
        .../here/share/lua/5.1/luacheck/main.lua:513: in function <.../here/share/lua/5.1/luacheck/main.lua:28>
        [C]: in function 'xpcall'
        .../here/share/lua/5.1/luacheck/main.lua:539: in main chunk
        [C]: in function 'require'
        ...e/here/lib/luarocks/rocks/luacheck/0.15.1-1/bin/luacheck:2: in main chunk
        [C]: at 0x004054c0
stack traceback:
        [C]: in function 'error'
        .../here/share/lua/5.1/luacheck/utils.lua:181: in function 'pcall'
        .../here/share/lua/5.1/luacheck/check.lua:209: in function 'func'
        .../here/share/lua/5.1/luacheck/utils.lua:261: in function 'map'
        .../here/share/lua/5.1/luacheck/main.lua:375: in function 'get_new_reports'
        .../here/share/lua/5.1/luacheck/main.lua:420: in function 'get_reports'
        .../here/share/lua/5.1/luacheck/main.lua:513: in function <.../here/share/lua/5.1/luacheck/main.lua:28>
        [C]: in function 'xpcall'
        .../here/share/lua/5.1/luacheck/main.lua:539: in main chunk
        [C]: in function 'require'
        ...e/here/lib/luarocks/rocks/luacheck/0.15.1-1/bin/luacheck:2: in main chunk
        [C]: at 0x004054c0
```

another error message:

```
Critical error: Luacheck 0.15.1 bug (please report at github.com/mpeterv/luacheck/issues):
/tmp/here/share/lua/5.1/luacheck/analyze.lua:257: attempt to index field 'node' (a nil value)
stack traceback:
        /tmp/here/share/lua/5.1/luacheck/utils.lua:162: in function '__index'
        /tmp/here/share/lua/5.1/luacheck/analyze.lua:257: in function 'check_unused_recursive_funcs'
        /tmp/here/share/lua/5.1/luacheck/analyze.lua:316: in function 'analyze'
        /tmp/here/share/lua/5.1/luacheck/check.lua:198: in function </tmp/here/share/lua/5.1/luacheck/check.lua:189>
        [C]: in function 'xpcall'
        /tmp/here/share/lua/5.1/luacheck/utils.lua:174: in function 'pcall'
        /tmp/here/share/lua/5.1/luacheck/check.lua:209: in function 'func'
        /tmp/here/share/lua/5.1/luacheck/utils.lua:261: in function 'map'
        /tmp/here/share/lua/5.1/luacheck/main.lua:375: in function 'get_new_reports'
        /tmp/here/share/lua/5.1/luacheck/main.lua:420: in function 'get_reports'
        /tmp/here/share/lua/5.1/luacheck/main.lua:513: in function </tmp/here/share/lua/5.1/luacheck/main.lua:28>
        [C]: in function 'xpcall'
        /tmp/here/share/lua/5.1/luacheck/main.lua:539: in main chunk
        [C]: in function 'require'
        /tmp/here/lib/luarocks/rocks/luacheck/0.15.1-1/bin/luacheck:2: in main chunk
        [C]: at 0x00405480
stack traceback:
        [C]: in function 'error'
        /tmp/here/share/lua/5.1/luacheck/utils.lua:181: in function 'pcall'
        /tmp/here/share/lua/5.1/luacheck/check.lua:209: in function 'func'
        /tmp/here/share/lua/5.1/luacheck/utils.lua:261: in function 'map'
        /tmp/here/share/lua/5.1/luacheck/main.lua:375: in function 'get_new_reports'
        /tmp/here/share/lua/5.1/luacheck/main.lua:420: in function 'get_reports'
        /tmp/here/share/lua/5.1/luacheck/main.lua:513: in function </tmp/here/share/lua/5.1/luacheck/main.lua:28>
        [C]: in function 'xpcall'
        /tmp/here/share/lua/5.1/luacheck/main.lua:539: in main chunk
        [C]: in function 'require'
        /tmp/here/lib/luarocks/rocks/luacheck/0.15.1-1/bin/luacheck:2: in main chunk
        [C]: at 0x00405480
```

**Comment by mraleph** [comment]
I looked briefly at the reproduction provided and by @starius and can see that we end up with a strange IR which ends with `IR_NOP` preceeeded by 2 `IR_RENAME`... Obviously snapshot restoration code can't deal with this and does not respect those renames - which causes it to restore some slots from incorrect place.

Unfortunately I don't have much time to dig further into this today. I will try to return to this tomorrow. 

If you are affected by this you can try to work around this bug by doing:

```c
diff --git a/src/lj_snap.c b/src/lj_snap.c
index 8638d9e..f67c783 100644
--- a/src/lj_snap.c
+++ b/src/lj_snap.c
@@ -310,7 +310,8 @@ static BloomFilter snap_renamefilter(GCtrace *T, SnapNo lim)
 {
   BloomFilter rfilt = 0;
   IRIns *ir;
-  for (ir = &T->ir[T->nins-1]; ir->o == IR_RENAME; ir--)
+  for (ir = &T->ir[T->nins-1]; ir->o == IR_NOP; ir--);
+  for (; ir->o == IR_RENAME; ir--)
     if (ir->op2 <= lim)
       bloomset(rfilt, ir->op1);
   return rfilt;
@@ -320,7 +321,8 @@ static BloomFilter snap_renamefilter(GCtrace *T, SnapNo lim)
 static RegSP snap_renameref(GCtrace *T, SnapNo lim, IRRef ref, RegSP rs)
 {
   IRIns *ir;
-  for (ir = &T->ir[T->nins-1]; ir->o == IR_RENAME; ir--)
+  for (ir = &T->ir[T->nins-1]; ir->o == IR_NOP; ir--);
+  for (; ir->o == IR_RENAME; ir--)
     if (ir->op1 == ref && ir->op2 <= lim)
       rs = ir->prev;
   return rs;
```

but this is just a nasty hack, not a real fix for the underlying problem. Underlying issue is that we end up with incorrectly sized IR when we finish assembling the trace and copy renames over - there is probably a logical error there which sets incorrect `nins` on the resulting trace.

**Comment by MikePall** [comment]
Really a duplicate of #183

Fixed by 01e4754962130dc85802a9738707d7fdb76c879c
